feat(parser): Add IoT registry events models

basvandriel opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** https://github.com/aws-powertools/powertools-lambda-python/issues/5891

## Summary

This implements the missing model for IoT registry events.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

basvandriel committed:
Add IoT crud event

basvandriel committed:
Fix typo

basvandriel committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

basvandriel committed:
work

basvandriel committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

basvandriel committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

basvandriel committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

basvandriel committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

leandrodamascena commented:
Hey @basvandriel, we need to address this comment before reviewing this PR: https://github.com/aws-powertools/powertools-lambda-python/issues/5891#issuecomment-2645728670.

Thanks

leandrodamascena committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

basvandriel committed:
Added all registry events

basvandriel committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

basvandriel commented:
Hi @leandrodamascena, I updated my branch to include all registry events, documented on [this page](https://docs.aws.amazon.com/iot/latest/developerguide/registry-events.html#). Please let me know if you would like to see additional changes, unit tests or any documentation. 

We could also add aliases such as `Field(alias='thingId')` values to be conform to the Python snake_case convention. However, I saw multiple models just having the names written in camelCase. Let me know! :-)

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    REMOVED = "REMOVED"
+
+
+class IoTCRUDEvent(IOTCoreRegistryEventBase):
```
```suggestion
class IoTCoreThingEvent(IoTCoreRegistryEventsBase):
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+from pydantic import BaseModel
+
+
+class IOTCoreRegistryEventBase(BaseModel):
```
```suggestion
class IoTCoreRegistryEventsBase(BaseModel):
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    attributes: Dict[str, Any]
+
+
+class IoTThingTypeCRUDEvent(IOTCoreRegistryEventBase):
```
```suggestion
class IoTCoreThingTypeEvent(IoTCoreRegistryEventsBase):
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    description: str
+
+
+class IoTThingTypeAssociationEvent(IOTCoreRegistryEventBase):
```
```suggestion
class IoTCoreThingTypeAssociationEvent(IoTCoreRegistryEventsBase):
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    groupId: str
+
+
+class IoTThingGroupCRUDEvent(IOTCoreRegistryEventBase):
```
```suggestion
class IoTCoreThingGroupEvent(IoTCoreRegistryEventsBase):
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    dynamicGroupMappingId: Optional[str]
+
+
+class IoTThingAddedToOrRemoveFromThingGroupEvent(IOTCoreRegistryEventBase):
```
```suggestion
class IoTCoreAddOrRemoveFromThingGroupEvent(IoTCoreRegistryEventsBase):
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    membershipId: str
+
+
+class IoTThingGroupAddedToOrDeletedFromThingGroupEvent(IOTCoreRegistryEventBase):
```
```suggestion
class IoTCoreAddOrDeleteFromThingGroupEvent(IoTCoreRegistryEventsBase):
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    timestamp: int
+
+
+class IoTCRUDEventOperation(str, enum.Enum):
+    CREATED = "CREATED"
+    UPDATED = "UPDATED"
+    DELETED = "DELETED"
+
+
+class AddRemoveOperation(str, enum.Enum):
+    ADDED = "ADDED"
+    REMOVED = "REMOVED"
```
```suggestion
EVENT_CRUD_OPERATION = Literal["CREATED", "UPDATED", "DELETED"]
EVENT_ADD_REMOVE_OPERATION = Literal["ADDED", "REMOVED"]
```

Please remove these `Enum` and transform them into a Literal type. It is easy to read and we don't need any type transformations to access the data. Also, move them to after the `import` section.

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+class IoTCRUDEvent(IOTCoreRegistryEventBase):
+    """The "Thing: created, updated, deleted" event"""
+
+    operation: IoTCRUDEventOperation
```
```suggestion
    operation: EVENT_CRUD_OPERATION
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+
+
+class IoTThingTypeCRUDEvent(IOTCoreRegistryEventBase):
+    operation: IoTCRUDEventOperation
```
```suggestion
    operation: EVENT_CRUD_OPERATION
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    Thing Group Created/Updated/Deleted
+    """
+
+    operation: IoTCRUDEventOperation
```
```suggestion
    operation: EVENT_CRUD_OPERATION
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    Thing Type Associated or Disassociated with a Thing
+    """
+
+    operation: AddRemoveOperation
```
```suggestion
    operation: EVENT_ADD_REMOVE_OPERATION
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    Thing Added to or Removed from a Thing Group
+    """
+
+    operation: AddRemoveOperation
```
```suggestion
    operation: EVENT_ADD_REMOVE_OPERATION
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+    Thing Group Added to or Deleted from a Thing Group
+    """
+
+    operation: AddRemoveOperation
```
```suggestion
    operation: EVENT_ADD_REMOVE_OPERATION
```

leandrodamascena commented on aws_lambda_powertools/utilities/parser/models/iot_registry_events.py:
```diff
@@ -0,0 +1,110 @@
+
+    operation: IoTCRUDEventOperation
+
+    thingId: str
+    thingName: str
+    versionNumber: int
+    thingTypeName: Optional[str]
+    billingGroupName: Optional[str]
+
+    attributes: Dict[str, Any]
```
```suggestion
    thingId: str
    accountId: str
    thingName: str
    versionNumber: int
    thingTypeName: Optional[str]
    billingGroupName: Optional[str]

    attributes: Dict[str, Any]
```

accountId is a key in this type of event.

leandrodamascena reviewed (changes requested):
Hi @basvandriel! Thank you so much for working on this PR and adding all the events! I have some feedback to make it more consistent and establish a naming standard.

Also, we need to include tests to verify that this is serializing as expected, you can copy this [AWS events](https://docs.aws.amazon.com/iot/latest/developerguide/registry-events.html) to test. In the documentation section https://docs.powertools.aws.dev/lambda/python/latest/utilities/parser/#built-in-models we need to add the new Models.

I also have a suggestion for the variable names, but I'd rather you correct the feedback I gave now and then we'll go to round two, okay?

leandrodamascena commented:
> We could also add aliases such as `Field(alias='thingId')` values to be conform to the Python snake_case convention. However, I saw multiple models just having the names written in camelCase. Let me know! :-)

Hey, please move on with this!

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Update aws_lambda_powertools/utilities/parser/models/iot_registry_events.py

Co-authored-by: Leandro Damascena <lcdama@amazon.pt>
Signed-off-by: Bas <contact@basvandriel.nl>

basvandriel committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

basvandriel committed:
work

basvandriel committed:
add tests

basvandriel committed:
rename tests

basvandriel committed:
rename test

leandrodamascena commented:
Hi @basvandriel! Please let me know when you need me to review it again, okay?

basvandriel commented:
> Hi @basvandriel! Please let me know when you need me to review it again, okay?

Hi @leandrodamascena, I made a misclick. I'll let you know!

basvandriel committed:
wokr

basvandriel committed:
Add documentation

basvandriel commented:
Hi @leandrodamascena, I've added the serialization tests and the documentation. I've tried to use the same format as the previous rows, e.g. `Lambda Event Source payload for IoT Core Thing created, updated, or deleted.`. Let me know like to see other changes.

basvandriel committed:
Remove stars

leandrodamascena committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

leandrodamascena committed:
Moving JSON events to events folder

leandrodamascena committed:
Moving JSON events to events folder

leandrodamascena reviewed (approved):
Hi @basvandriel! Thanks for working on this PR and resolving all the comments! This is really great work and contribution!

I just pushed a commit with a few changes:

1 - Added docstrings to all classes
2 - Moved JSON events to the events folder because people often come to our repository to copy these events as a source of truth.
3 - I have a feeling that `rootToParentThingGroups` is a list with arbitrary fields and not only`arnid`. So I converted it to a list of Dicts to have a guardrail.
4 - I added the even_type field with literal values ​​to the corresponding classes.

I'm approving this PR and we'll merge it next week. We will release it in the next release of Powertools, which is expected on 2/25.

leandrodamascena committed:
Merge branch 'develop' into feature/iot-core-crud-event-types

basvandriel commented:
Thanks @leandrodamascena. Lets stay in touch - I'd love to work on the project some more.

leandrodamascena commented:
> Thanks @leandrodamascena. Lets stay in touch - I'd love to work on the project some more.

That's great to know! How about you add these events to our Event Source Data Class utility? The main difference is that in this utility we don't use Pydantic to do strict validation, but instead we convert the event to an object to help developers access properties like a python object. Here are some examples: https://github.com/aws-powertools/powertools-lambda-python/tree/develop/aws_lambda_powertools/utilities/data_classes.

Please let me know what you think and if you agree, feel free to open an issue and send a PR.

leandrodamascena committed:
Merge branch 'develop' into feature/iot-core-crud-event-types
